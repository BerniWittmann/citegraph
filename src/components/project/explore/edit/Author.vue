<template>
  <div class="project-explore-single-author-edit">
    <v-card class="pa-4">
      <v-card-title class="headline">
        <h2 class="display-5" v-once>{{ author.firstName }} {{ author.lastName }}</h2>
      </v-card-title>
      <v-container>
        <v-form ref="form" @submit.prevent="submit" class="mt-12" v-model="valid">
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.first_name')"
                v-model="firstName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.last_name')"
                v-model="lastName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row class="my-10" justify="space-between">
            <v-col cols="auto">
              <v-btn rounded large color="warning" @click="back">
                <v-icon>mdi-menu-left</v-icon>
                {{ $t('project.explore.edit.back') }}
              </v-btn>
            </v-col>

            <v-col cols="auto">
              <v-btn
                rounded large
                :disabled="!valid"
                :loading="loading"
                color="success"
                type="submit"
              >{{ $t('project.explore.edit.submit') }}
              </v-btn>
            </v-col>
          </v-row>
        </v-form>
      </v-container>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Author from '@/models/paperEntities/author'

@Component
export default class AuthorEditPage extends Vue {
  valid: boolean = true
  loading: boolean = false
  firstName: string = ''
  lastName: string = ''

  nameRules: Array<Function> = [
    (v: string | undefined) => !!v || this.$root.$t('project.explore.edit.author.name_required'),
    (v: string | undefined) => (v && v.trim().length > 0) || this.$root.$t('project.explore.edit.author.name_not_empty')
  ]

  urlRules: Array<Function> = []

  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  beforeMount (): void {
    this.firstName = this.author.firstName
    this.lastName = this.author.lastName
  }

  async submit (): Promise<void> {
    if (!this.valid) return
    this.loading = true
    await this.$store.dispatch('paperEntities/updateEntity', {
      params: {
        projectId: this.$route.params.projectId,
        entityType: this.$route.params.queryByType,
        id: this.$route.params.entityId
      },
      data: {
        firstName: this.firstName,
        lastName: this.lastName
      }
    })
    this.loading = false
    this.$router.back()
  }

  back (): void {
    this.$router.back()
  }
}
</script>
