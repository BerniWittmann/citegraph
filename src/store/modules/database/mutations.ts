import { DatabaseState } from './types'
import { MutationTree } from 'vuex'

export const mutations: MutationTree<DatabaseState> = {}
