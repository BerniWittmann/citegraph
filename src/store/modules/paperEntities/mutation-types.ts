export const SET_ENTITIES = 'SET_ENTITIES'
export const SET_ENTITY_TYPE = 'SET_ENTITY_TYPE'
export const SET_ENTITY_COUNT = 'SET_ENTITY_COUNT'
