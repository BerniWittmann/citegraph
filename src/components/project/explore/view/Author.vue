<template>
  <div class="project-explore-single-author">
    <layout>
      <template v-slot:title>
        <v-avatar color="primary" class="mr-5" size="62">
          <b class="white--text">{{ initials }}</b>
        </v-avatar>
        <h2 class="display-5">{{ author.firstName }} {{ author.lastName }}</h2>
      </template>
      <v-container>
        <v-row>
          <v-col :cols="3">
            <v-list>
              <v-subheader class="title">{{ $t('project.explore.view.author.metrics') }}</v-subheader>
              <v-list-item>
                <v-list-item-content>{{ $t('project.explore.view.author.count_records') }}:</v-list-item-content>
                <v-list-item-content class="align-end">{{ author.countRecords }}</v-list-item-content>
              </v-list-item>
            </v-list>
          </v-col>
        </v-row>
      </v-container>
    </layout>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import PaperEntityLayout from '@/layouts/PaperEntity.vue'
import Author from '@/models/paperEntities/author'

@Component({
  components: {
    Layout: PaperEntityLayout
  }
})
export default class AuthorViewPage extends Vue {
  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  get initials (): string {
    return (this.author.firstName.substring(0, 1) + this.author.lastName.substring(0, 1)).toUpperCase()
  }
}
</script>
