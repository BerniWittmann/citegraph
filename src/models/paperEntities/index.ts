import Author from './author'
import Conference from './conference'
import Country from './country'
import Institution from './institution'
import Publication from './publication'
import Publisher from './publisher'
import Record from './record'
import PaperEntity, { PaperEntityTableColumn, PaperEntityFields } from './base'

export const entities = [
  Record,
  Author,
  Institution,
  Country,
  Publication,
  Publisher,
  Conference
]

export const entityKeysMap = {
  [Record.key]: Record,
  [Author.key]: Author,
  [Institution.key]: Institution,
  [Country.key]: Country,
  [Publication.key]: Publication,
  [Publisher.key]: Publisher,
  [Conference.key]: Conference
}

export const entityKeys: Array<string> = entities.map((entity) => entity.key)

export type PaperEntityTableColumn = PaperEntityTableColumn

export type PaperEntity = PaperEntity

export type PaperEntityFields = PaperEntityFields

export default {
  Author,
  Conference,
  Country,
  Institution,
  Publication,
  Publisher,
  Record
}
