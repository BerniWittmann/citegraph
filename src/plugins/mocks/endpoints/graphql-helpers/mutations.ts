import { PaperEntityFields } from '@/models/paperEntities'
import { QueryConfiguration } from '@/plugins/mocks/endpoints/graphql-helpers/queries'
import db from '../../db'

type MutationArguments = {
  id?: string
}

function parseQuery (query: string): MutationArguments {
  const idString = query.match(/id: [a-zA-Z0-9]+/)
  const id = idString ? idString[0].toString().substring(4) : undefined
  return {
    id
  }
}

function parseData (query: string): Partial<PaperEntityFields> {
  // extract update data
  const result: Record<string, any> = {}
  const items = query.match(/\w+: ".*"/g) || []
  items.forEach((item: string) => {
    const key = item.split(': ')[0]
    const val = item.split(': ')[1]
    result[key] = val.substring(1, val.length - 1).replace(/\/"/g, '')
  })
  return result
}

function updateOne (queryParams: MutationArguments, data: Partial<PaperEntityFields>, configuration: QueryConfiguration) {
  let entity = configuration.collection.findOne({ id: queryParams.id })
  if (!entity) {
    throw new Error('Entity not found')
  }
  for (let key in data) {
    // @ts-ignore
    entity[key] = data[key]
  }
  return configuration.collection.update(entity)
}

export function createMutationResponse (query: string, configuration: QueryConfiguration) {
  const queryParams = parseQuery(query)
  if (!queryParams.id) {
    return [400, 'No id given']
  }
  try {
    const response = updateOne(queryParams, parseData(query), configuration)
    db.saveDatabase()
    return [200, {
      data: {
        [configuration.queryName]: {
          ...response
        }
      }
    }]
  } catch (err) {
    return [404, err]
  }
}
