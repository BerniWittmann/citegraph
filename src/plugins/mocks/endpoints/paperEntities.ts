import db from '../db'
import MockAdapter from 'axios-mock-adapter'
import { RecordFields } from '@/models/paperEntities/record'
import { setupMocks as setupGraphQLMocks } from './graphql-helpers/endpoint'
import { QueryConfiguration } from './graphql-helpers/queries'
import { AuthorFields } from '@/models/paperEntities/author'
import { CountryFields } from '@/models/paperEntities/country'

const records = db.getCollection('records')
const authors = db.getCollection('authors')
const countries = db.getCollection('countries')

const queryResponseConfiguration: Record<string, QueryConfiguration> = {
  'Records': {
    schemaName: 'record',
    queryName: 'Records',
    collection: records,
    filterFunction: recordsFilter
  },
  'Authors': {
    schemaName: 'author',
    queryName: 'Authors',
    collection: authors,
    filterFunction: authorsFilter
  },
  'Countries': {
    schemaName: 'country',
    queryName: 'Countries',
    collection: countries,
    filterFunction: countriesFilter
  }
}

function recordsFilter (obj: RecordFields, filter: string) {
  return obj.title.toLowerCase().includes(filter) ||
    obj.year.toString().includes(filter) ||
    obj.keywords.map(keyword => keyword.toLowerCase()).join('').includes(filter) ||
    obj.authors.map(author => (author.firstName + ' ' + author.lastName).toLowerCase()).join(';').includes(filter)
}

function authorsFilter (obj: AuthorFields, filter: string) {
  return obj.firstName.toLowerCase().includes(filter) ||
    obj.lastName.toLowerCase().includes(filter)
}

function countriesFilter (obj: CountryFields, filter: string) {
  return obj.name.toLowerCase().includes(filter)
}

export function setupMocks (mock: MockAdapter) {
  setupGraphQLMocks(mock, queryResponseConfiguration)
}
