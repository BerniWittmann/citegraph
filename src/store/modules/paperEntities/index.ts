import { Module } from 'vuex'
import { PaperEntitiesState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const paperEntities: Module<PaperEntitiesState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
