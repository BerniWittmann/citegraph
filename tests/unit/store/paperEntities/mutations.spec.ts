import { mutations } from '@/store/modules/paperEntities/mutations'
import * as mutationTypes from '@/store/modules/paperEntities/mutation-types'
import { PaperEntitiesState } from '@/store/modules/paperEntities/types'
import { PaperEntity } from '@/models/paperEntities'
import Author from '@/models/paperEntities/author'

describe('store/modules/database/mutations', () => {
  it('provides the mutations', () => {
    expect(mutations).toMatchSnapshot()
  })

  describe('SET_ENTITIES', () => {
    it('saves the entities to the store', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: '',
        activeEntity: undefined
      }
      const entities: Array<PaperEntity> = [
        new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
        new Author({ id: '2', firstName: 'Maria', lastName: 'Maier', countRecords: 24 })
      ]

      mutations[mutationTypes.SET_ENTITIES](state, entities)

      expect(state.entities).toEqual(entities)
    })

    it('overrides existing entities', () => {
      const state: PaperEntitiesState = {
        entities: [
          new Author({ id: '42', firstName: 'Other', lastName: 'Author', countRecords: 12 }),
          new Author({ id: '99', firstName: 'hunter', lastName: 'Two', countRecords: 24 })
        ],
        entityCount: -1,
        entityType: '',
        activeEntity: undefined
      }
      const entities: Array<PaperEntity> = [
        new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
        new Author({ id: '2', firstName: 'Maria', lastName: 'Maier', countRecords: 24 })
      ]

      mutations[mutationTypes.SET_ENTITIES](state, entities)

      expect(state.entities).toEqual(entities)
    })
  })

  describe('SET_ENTITY_TYPE', () => {
    it('saves the entity type to the store', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: '',
        activeEntity: undefined
      }
      const type = 'record'

      mutations[mutationTypes.SET_ENTITY_TYPE](state, type)

      expect(state.entityType).toEqual(type)
    })

    it('overrides existing entity type', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: 'author',
        activeEntity: undefined
      }
      const type = 'record'

      mutations[mutationTypes.SET_ENTITY_TYPE](state, type)

      expect(state.entityType).toEqual(type)
    })
  })

  describe('SET_ENTITY_COUNT', () => {
    it('saves the entity count to the store', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: '',
        activeEntity: undefined
      }
      const count = 12

      mutations[mutationTypes.SET_ENTITY_COUNT](state, count)

      expect(state.entityCount).toEqual(count)
    })

    it('overrides existing entity count', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: 99,
        entityType: 'author',
        activeEntity: undefined
      }

      const count = 12

      mutations[mutationTypes.SET_ENTITY_COUNT](state, count)

      expect(state.entityCount).toEqual(count)
    })
  })

  describe('SET_ACTIVE_ENTITY', () => {
    it('saves the entity to the store', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: '',
        activeEntity: undefined
      }
      const entity = new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })

      mutations[mutationTypes.SET_ACTIVE_ENTITY](state, entity)

      expect(state.activeEntity).toEqual(entity)
    })

    it('overrides existing entity ', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: '',
        activeEntity: new Author({ id: '2', firstName: 'Other', lastName: 'User', countRecords: 12 })
      }

      const entity = new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })

      mutations[mutationTypes.SET_ACTIVE_ENTITY](state, entity)

      expect(state.activeEntity).toEqual(entity)
    })

    it('can unset an existing entity ', () => {
      const state: PaperEntitiesState = {
        entities: [],
        entityCount: -1,
        entityType: '',
        activeEntity: new Author({ id: '2', firstName: 'Other', lastName: 'User', countRecords: 12 })
      }

      mutations[mutationTypes.SET_ACTIVE_ENTITY](state, undefined)

      expect(state.activeEntity).toEqual(undefined)
    })
  })

  describe('UPDATE_ENTITY', () => {
    it('updates the correct entity', () => {
      const state: PaperEntitiesState = {
        entities: [
          new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '2', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '3', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
        ],
        entityCount: -1,
        entityType: '',
        activeEntity: undefined
      }
      const entity = new Author({ id: '2', firstName: 'Neuer', lastName: 'Name', countRecords: 10 })

      mutations[mutationTypes.UPDATE_ENTITY](state, entity)

      expect(state.entities.length).toEqual(3)
      expect(state.entities).toContain(entity)
    })

    it('does nothing if entity does not exist', () => {
      const state: PaperEntitiesState = {
        entities: [
          new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '2', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '3', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
        ],
        entityCount: -1,
        entityType: '',
        activeEntity: new Author({ id: '3', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
      }
      const entity = new Author({ id: '99', firstName: 'Neuer', lastName: 'Name', countRecords: 10 })

      mutations[mutationTypes.UPDATE_ENTITY](state, entity)

      expect(state.entities.length).toEqual(3)
      expect(state.entities).not.toContain(entity)
      expect(state.activeEntity).not.toEqual(entity)
    })

    it('updates the active Entity', () => {
      const state: PaperEntitiesState = {
        entities: [
          new Author({ id: '1', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '2', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 }),
          new Author({ id: '3', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
        ],
        entityCount: -1,
        entityType: '',
        activeEntity: new Author({ id: '2', firstName: 'Max', lastName: 'Mustermann', countRecords: 12 })
      }
      const entity = new Author({ id: '2', firstName: 'Neuer', lastName: 'Name', countRecords: 10 })

      mutations[mutationTypes.UPDATE_ENTITY](state, entity)

      expect(state.activeEntity).toEqual(entity)
    })
  })
})
