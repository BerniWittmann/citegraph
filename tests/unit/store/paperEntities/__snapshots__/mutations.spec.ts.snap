// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/database/mutations provides the mutations 1`] = `
Object {
  "SET_ACTIVE_ENTITY": [Function],
  "SET_ENTITIES": [Function],
  "SET_ENTITY_COUNT": [Function],
  "SET_ENTITY_TYPE": [Function],
}
`;
